### ISSUE TITLE
Indexed decision design flaw with NameUsageWrapper 

### ISSUE THREAD

**Issue by mdoering** [Sp2000/colplus-backend#569]
The indexed name usages make use of a NameUsageWrapper class that contains a usage together with a bunch of other things, e.g. the entire classification, vernacular names and a decisionKey.

This was all fine when we only had a single catalogue, as any usage could only ever have a single decision. But [since we allow decisions on the same usage for different catalogue projects](https://github.com/Sp2000/colplus-backend/issues/484), we need to index them all in ES.

As we likely want to filter decisions by their datasetKey, i.e. the catalogue they belong to (not the subjectDatasetKey), it is not enough to store a list of keys.

We must either store the list of entire decisions (so the user interface can filter and show them directly without a lookup) or at least store a map of the catalogues datasetKey -> decisionKey. I would think we'd also want to filter by the decision mode in the future (e.g. reviewed decisions only), so we should have a map with objects as their value. Or two maps both keyed on the catalogue and having the decisionKey and the decision mode as values.

**Comment by mdoering** [comment]
Proposal:

- add a new `NameUsageWrapper.decisions` list of `SimpleDecision` objects
   - this is what ES needs to index and offer ways to search for MODE and datasetKey
- remove `NameUsageWrapper.decisionKey`
- replace `NameUsageSearchParameter.DECISION_KEY` with `NameUsageSearchParameter.DECISION_DATASET_KEY`: this takes the datasetKey of the catalogue to filter by. Only usages with a decision from that catalogue are returned
- add `NameUsageSearchParameter.DECISION_NO_DATASET_KEY`: in the absence of a generic NOT request operator we add another search parameter that takes a catalogues datasetKey and returns all usages that do NOT have a decision in the given catalogue.
- add `NameUsageSearchRequest.catalogueKey` parameter to prune the returned usage objects for data relevant to the given catalogue only. In particular this means the `decisions` list returned from the search just contains a single decision at max so the UI can quickly determine if a decision exists for a given usage and does not have to ignore decisions from other catalogues

```
public class SimpleDecision {
  private Integer key;
  private Integer datasetKey; // the catalogues datasetKey
  private EditorialDecision.Mode mode;
}
```

**Comment by thomasstjerne** [comment]
I guess that `NameUsageSearchParameter.DECISION_DATASET_KEY` would translate into `/dataset/1700/nameusage/search?decisionDatasetKey=3` in my world, correct?

**Comment by thomasstjerne** [comment]
And how do we query for usages that has no decision in a given dataset?

**Comment by thomasstjerne** [comment]
> And how do we query for usages that has no decision in a given dataset?

Ah, sorry forget that, DECISION_MODE = NULL it is

**Comment by mdoering** [comment]
Yes, decisionDatasetKey query parameter would be the translation.
The most basic endpoint is not scoped to a dataset at all:

`/nameusage/search?decisionDatasetKey=3`

We do also expose the exact same search within a dataset scope, so the path parameter translates into a query param if you like and the following are identical:
`/nameusage/search?datasetKey=1010&decisionDatasetKey=3&q=Abies`
`/dataset/1010/nameusage/search?decisionDatasetKey=3&q=Abies`

We could expose it yet again under another path as you suggest.
I wonder if this is really a good idea though and maybe we should also remove the dataset scoped search. 

In general would you favor `/catalogue` over `/assembly` ?

**Comment by mdoering** [comment]
Left todo:

- [ ] mybatis implementation to return the decisions lists for nu wrappers
- [ ] elasticsearch implementation to store the relevant bits in the index and remove the former decisionKey
- [ ] update elasticsearch tests, many of which have been ignored now and contain many `//TODO` remarks

**Comment by thomasstjerne** [comment]
> We could expose it yet again under another path as you suggest.
I wonder if this is really a good idea though and maybe we should also remove the dataset scoped search.

It is not a big deal, its just that the frontend code seems cleaner when the scope is defined in path params. For now, ignore `/catalogue/3/dataset/1700/nameusage/search` suggestion, but please keep the dataset path param until I have finished the large rework of catalogue / dataset scope.
Then I will switch to the datasetKey query param and let you know when the frond doesn´t depend on the path param any longer

**Comment by ayco-at-naturalis** [comment]
Yes SimpleDecision sounds fine. Note that do get another nested object then (so a multiplication of sub-documents), but the document structure is still agreeably flat

**Comment by ayco-at-naturalis** [comment]
(The thing I most wonder about is whether the SQL to get all this out of Postgres in one shot isn't getting totally out of hand)

**Comment by mdoering** [comment]
sql is fine, we do get by id for each document these days

**Comment by ayco-at-naturalis** [comment]
Note: as per our discussions, currently only the decision's mode and catalogue key is stored in the index; the decision key is ignored. I cannot remember though why I thought this came out of our discussions. @thomasstjerne @mdoering Is this correct? Does Thomas not need the decision key to pull in the full-blown decision?

**Comment by ayco-at-naturalis** [comment]
Closed in 65d8816

**Comment by mdoering** [comment]
> Note: as per our discussions, currently only the decision's mode and catalogue key is stored in the index; the decision key is ignored. I cannot remember though why I thought this came out of our discussions. @thomasstjerne @mdoering Is this correct? Does Thomas not need the decision key to pull in the full-blown decision?

Correct that we only need to search on mode and catalogue key. The decision(s) returned should obviously be the full ones submitted to ES just as we do for the NameUsageWrapper, i.e. they must include the decisionKey:

```
public class SimpleDecision {
  private Integer key;
  private Integer datasetKey;
  private EditorialDecision.Mode mode;

```
